File: src/preview_system.cpp
#include "preview_system.hpp"
#include "chunk_system.hpp"
#include "command.hpp"
#include "room.hpp"
#include "util/mutex.hpp"
#include <cassert>

// static const char *LOG_PREVIEW_SYSTEM = "PreviewSystem";
static const char *LOG_PREVIEW_SYSTEM_LAYER = "PreviewSystemLayer";

PreviewSystemLayer::PreviewSystemLayer(PreviewSystem *system)
		: system(system) {
}

void PreviewSystemLayer::addToQueue(Int2 coords) {
	// Check if already added to queue (reverse iterator)
	for(auto it = update_queue.rbegin(); it != update_queue.rend(); it++) {
		if(*it == coords)
			return;
	}

	// Add to queue
	update_queue.push_back(coords);
}

bool PreviewSystemLayer::processOneBlock() {
	if(update_queue.empty())
		return false;

	auto position = update_queue.front();
	update_queue.pop_front();

	Int2 topleft = {position.x * 2, position.y * 2};
	Int2 topright = {position.x * 2 + 1, position.y * 2};
	Int2 bottomleft = {position.x * 2, position.y * 2 + 1};
	Int2 bottomright = {position.x * 2 + 1, position.y * 2 + 1};

	SharedVector<u8> compressed_topleft, compressed_topright, compressed_bottomleft, compressed_bottomright; // Compressed data

	// Lock database and fetch compressed data
	auto &database = system->room->database;
	database.lock();
	if(zoom == 1) { // Real chunks underneath
		compressed_topleft = database.chunkLoadData(topleft).data;
		compressed_topright = database.chunkLoadData(topright).data;
		compressed_bottomleft = database.chunkLoadData(bottomleft).data;
		compressed_bottomright = database.chunkLoadData(bottomright).data;
	} else {
		compressed_topleft = database.previewLoadData(topleft, zoom - 1).data;
		compressed_topright = database.previewLoadData(topright, zoom - 1).data;
		compressed_bottomleft = database.previewLoadData(bottomleft, zoom - 1).data;
		compressed_bottomright = database.previewLoadData(bottomright, zoom - 1).data;
	}

	// Unlock database
	database.unlock();

	const auto chunk_size = ChunkSystem::getChunkSize();

	// Decompress data
	auto decompress = [chunk_size](SharedVector<u8> &in) -> SharedVector<u8> {
		if(!in) return {};
		auto out = createSharedVector<u8>(chunk_size * chunk_size * 3);
		auto ret = decompressLZ4(in->data(), in->size(), out->data(), out->size());
		assert(ret >= 0);
		(void)ret;
		return out;
	};

	auto data_topleft = decompress(compressed_topleft);
	auto data_topright = decompress(compressed_topright);
	auto data_bottomleft = decompress(compressed_bottomleft);
	auto data_bottomright = decompress(compressed_bottomright);

	// Fuse 2x2 chunks into one image
	const auto image_size = chunk_size * 2;
	uniqdata<u8> rgb(image_size * image_size /* 512² */ * 3 /*RGB*/);
	memset(rgb.data(), 255, rgb.size());

	auto fillImage = [&](SharedVector<u8> &data, u8 x, u8 y) {
		if(!data) return;

		u32 offset_x = chunk_size * x;
		u32 offset_y = chunk_size * y;

		const auto *rgb_in = data->data();
		u32 pitch_in = chunk_size * 3;
		u32 pitch_out = image_size * 3;

		for(u32 local_y = 0; local_y < chunk_size; local_y++) {
			for(u32 local_x = 0; local_x < chunk_size; local_x++) {
				u32 target_x = offset_x + local_x;
				u32 target_y = offset_y + local_y;

				u32 offset_in = local_y * pitch_in + local_x * 3;
				u32 offset_out = target_y * pitch_out + target_x * 3;

				rgb[offset_out + 0] = rgb_in[offset_in + 0];
				rgb[offset_out + 1] = rgb_in[offset_in + 1];
				rgb[offset_out + 2] = rgb_in[offset_in + 2];
			}
		}
	};

	// Blit images
	fillImage(data_topleft, 0, 0);
	fillImage(data_topright, 1, 0);
	fillImage(data_bottomleft, 0, 1);
	fillImage(data_bottomright, 1, 1);

	// Downscale image
	uniqdata<u8> downscaled(chunk_size * chunk_size * 3);
	const u32 downscaled_pitch = chunk_size * 3;
	const u32 image_pitch = image_size * 3;
	for(u32 y = 0; y < chunk_size; y++) {
		for(u32 x = 0; x < chunk_size; x++) {
			u32 in_x = x * 2;
			u32 in_y = y * 2;

			auto *out = &downscaled[y * downscaled_pitch + x * 3];

			auto performChannel = [&](u8 channel) {
				out[channel] =
						((u32)rgb[(in_y + 0) * image_pitch + (in_x + 0) * 3 + channel] +
						 (u32)rgb[(in_y + 1) * image_pitch + (in_x + 0) * 3 + channel] +
						 (u32)rgb[(in_y + 0) * image_pitch + (in_x + 1) * 3 + channel] +
						 (u32)rgb[(in_y + 1) * image_pitch + (in_x + 1) * 3 + channel]) /
						4;
			};

			performChannel(0); // Red
			performChannel(1); // Green
			performChannel(2); // Blue
		}
	}

	// Compress downscaled image
	auto compressed = compressLZ4(downscaled.data(), downscaled.size());

	// Write result, Lock database again
	database.lock();
	database.previewSaveData(position, zoom, compressed->data(), compressed->size());
	// Unlock database
	database.unlock();

	if(upper_layer) {
		s32 x = position.x / 2;
		s32 y = position.y / 2;
		if(position.x < 0) x--;
		if(position.y < 0) y--;
		upper_layer->addToQueue({x, y});
	}

	system->room->log(LOG_PREVIEW_SYSTEM_LAYER, "Processed block at zoom %u (%u remaining)", zoom, (u32)update_queue.size());
	return true;
}

PreviewSystem::PreviewSystem(Room *room)
		: room(room) {
	// Generate layers
	layers.resize(getLayerCount(), this);
	for(u32 i = 0; i < getLayerCount(); i++) {
		auto &layer = layers[i];
		layer.zoom = layerIndexToZoom(i);
		if(i < getLayerCount() - 1)
			layer.upper_layer = &layers[i + 1];
	}
}

void PreviewSystem::tick() {
	{
		LockGuard lock(mtx_access);
		for(auto &pos : update_queue_cache) {
			layers[0].addToQueue(pos);
		}
		update_queue_cache.clear();
	}

	for(auto &layer : layers) {
		if(!layer.processOneBlock())
			continue;
		break;
	}

	queue.process();
}

void PreviewSystem::addToQueueFront(Int2 coords) {
	LockGuard lock(mtx_access);
	update_queue_cache.push_back(coords);
}

PreviewSystem::~PreviewSystem() {
}

u32 PreviewSystem::getLayerCount() const {
	return 4;
}

SharedVector<u8> PreviewSystem::requestData(s32 preview_x, s32 preview_y, u8 zoom) {
	auto &database = room->database;
	database.lock();
	auto data = database.previewLoadData({preview_x, preview_y}, zoom).data;
	database.unlock();
	return data;
}

File: src/preview_system.rs
#![allow(dead_code)]

use std::sync::{Arc, Weak};

use anyhow::anyhow;
use glam::IVec2;
use tokio::sync::{Mutex, Notify};

use crate::{
	compression,
	database::{ChunkDatabaseRecord, Database, DatabaseFunc, PreviewDatabaseRecord},
	event_queue::EventQueue,
	limits::{self, CHUNK_SIZE_PX},
};

#[derive(Default, Clone)]
pub struct PreviewSystemLayer {
	zoom: u8,
	update_queue: Vec<IVec2>,
}

fn extract_chunk_record(record: Option<ChunkDatabaseRecord>) -> Vec<u8> {
	if let Some(record) = record {
		record.data
	} else {
		Vec::new()
	}
}

fn extract_preview_record(record: Option<PreviewDatabaseRecord>) -> Vec<u8> {
	if let Some(record) = record {
		record.data
	} else {
		Vec::new()
	}
}

fn decompress_vec_lz4(compressed: &[u8]) -> anyhow::Result<Vec<u8>> {
	if compressed.is_empty() {
		return Ok(Vec::new());
	}
	compression::decompress_lz4(compressed, (CHUNK_SIZE_PX * CHUNK_SIZE_PX * 4) as usize)
		.map_or_else(|| Err(anyhow!("Failed to decompress LZ4 data")), Ok)
}

fn fill_image(out_data: &mut [u8], in_data: &[u8], x: u32, y: u32) {
	if in_data.is_empty() {
		return; // blank
	}
	let offset_x = CHUNK_SIZE_PX * x;
	let offset_y = CHUNK_SIZE_PX * y;

	let image_size = CHUNK_SIZE_PX * 2;
	let pitch_in = CHUNK_SIZE_PX * 4;
	let pitch_out = image_size * 4;

	for local_y in 0..CHUNK_SIZE_PX {
		for local_x in 0..CHUNK_SIZE_PX {
			let target_x = offset_x + local_x;
			let target_y = offset_y + local_y;

			let offset_in = local_y * pitch_in + local_x * 4;
			let offset_out = target_y * pitch_out + target_x * 4;

			out_data[(offset_out) as usize] = in_data[(offset_in) as usize];
			out_data[(offset_out + 1) as usize] = in_data[(offset_in + 1) as usize];
			out_data[(offset_out + 2) as usize] = in_data[(offset_in + 2) as usize];
			out_data[(offset_out + 3) as usize] = in_data[(offset_in + 3) as usize];
		}
	}
}

struct PreviewProcessResult {
	has_more: bool,
	upper_pos: IVec2,
}

fn decompress_and_blit(
	compressed: &[u8],
	x: u32,
	y: u32,
	out_rgba: &mut [u8],
) -> anyhow::Result<()> {
	let data = decompress_vec_lz4(compressed)?;
	fill_image(out_rgba, &data, x, y);
	Ok(())
}

impl PreviewSystemLayer {
	pub const fn new(zoom: u8) -> Self {
		Self {
			zoom,
			update_queue: Vec::new(),
		}
	}

	pub fn add_to_queue(&mut self, coords: IVec2) {
		// Check if already added to queue (reverse iterator)
		for it in self.update_queue.iter().rev() {
			if *it == coords {
				return; // Already added to the queue
			}
		}

		self.update_queue.push(coords);
	}

	#[allow(clippy::too_many_lines)]
	async fn process_one_block(
		database: &Arc<Mutex<Database>>,
		update_queue: &mut Vec<IVec2>,
		zoom: u8,
	) -> anyhow::Result<PreviewProcessResult> {
		struct Quad {
			topleft: Vec<u8>,
			topright: Vec<u8>,
			bottomleft: Vec<u8>,
			bottomright: Vec<u8>,
		}

		let Some(position) = update_queue.pop() else {
			// Nothing to do
			return Ok(PreviewProcessResult {
				has_more: false,
				upper_pos: IVec2::ZERO,
			});
		};

		// Fuse 2x2 chunks into one preview image
		let topleft = IVec2::new(position.x * 2, position.y * 2);
		let topright = IVec2::new(position.x * 2 + 1, position.y * 2);
		let bottomleft = IVec2::new(position.x * 2, position.y * 2 + 1);
		let bottomright = IVec2::new(position.x * 2 + 1, position.y * 2 + 1);

		let compressed = if zoom == 1 {
			// Load real chunks data underneath
			Quad {
				topleft: extract_chunk_record(DatabaseFunc::chunk_load_data(database, topleft).await?),
				topright: extract_chunk_record(DatabaseFunc::chunk_load_data(database, topright).await?),
				bottomleft: extract_chunk_record(
					DatabaseFunc::chunk_load_data(database, bottomleft).await?,
				),
				bottomright: extract_chunk_record(
					DatabaseFunc::chunk_load_data(database, bottomright).await?,
				),
			}
		} else {
			// Load preview system layer chunks
			Quad {
				topleft: extract_preview_record(
					DatabaseFunc::preview_load_data(database, topleft, zoom - 1).await?,
				),
				topright: extract_preview_record(
					DatabaseFunc::preview_load_data(database, topright, zoom - 1).await?,
				),
				bottomleft: extract_preview_record(
					DatabaseFunc::preview_load_data(database, bottomleft, zoom - 1).await?,
				),
				bottomright: extract_preview_record(
					DatabaseFunc::preview_load_data(database, bottomright, zoom - 1).await?,
				),
			}
		};

		let compressed = tokio::task::spawn_blocking(move || -> anyhow::Result<Vec<u8>> {
			// Allocate preview chunk data
			let image_size = CHUNK_SIZE_PX * 2;
			let mut rgba: Vec<u8> = vec![0; (image_size * image_size /* 512² */ * 4/*RGBA*/) as usize];

			// Decompress data
			decompress_and_blit(&compressed.topleft, 0, 0, &mut rgba)?;
			decompress_and_blit(&compressed.topright, 1, 0, &mut rgba)?;
			decompress_and_blit(&compressed.bottomleft, 0, 1, &mut rgba)?;
			decompress_and_blit(&compressed.bottomright, 1, 1, &mut rgba)?;

			// Downscale image
			let mut downscaled: Vec<u8> = vec![0; (CHUNK_SIZE_PX * CHUNK_SIZE_PX * 4) as usize];

			let downscaled_pitch = CHUNK_SIZE_PX * 4;
			let image_pitch = image_size * 4;
			for y in 0..CHUNK_SIZE_PX {
				for x in 0..CHUNK_SIZE_PX {
					let in_x = x * 2;
					let in_y = y * 2;

					let mut perform_channel = |channel: u32| {
						// Same four pixels
						let result = (u32::from(rgba[((in_y) * image_pitch + (in_x) * 4 + channel) as usize])
							+ u32::from(rgba[((in_y + 1) * image_pitch + (in_x) * 4 + channel) as usize])
							+ u32::from(rgba[((in_y) * image_pitch + (in_x + 1) * 4 + channel) as usize])
							+ u32::from(rgba[((in_y + 1) * image_pitch + (in_x + 1) * 4 + channel) as usize]))
							/ 4;

						// Save sampled pixel result for specific channel
						downscaled[(y * downscaled_pitch + x * 4 + channel) as usize] = result as u8;
					};

					perform_channel(0); // Red
					perform_channel(1); // Green
					perform_channel(2); // Blue
					perform_channel(3); // Alpha
				}
			}

			// Compress downscaled image
			let compressed = compression::compress_lz4(&downscaled);

			Ok(compressed)
		})
		.await??;

		DatabaseFunc::preview_save_data(database, position, zoom, compressed).await?;

		let upper_pos = IVec2::new(
			/* X */
			if position.x >= 0 {
				position.x / 2
			} else {
				(position.x - 1) / 2
			},
			/* Y */
			if position.y >= 0 {
				position.y / 2
			} else {
				(position.y - 1) / 2
			},
		);

		log::trace!(
			"Processed block at {}x{}, zoom {} ({} remaining)",
			position.x,
			position.y,
			zoom,
			update_queue.len()
		);

		Ok(PreviewProcessResult {
			has_more: true,
			upper_pos,
		})
	}
}

pub type PreviewSystemQueuedChunks = EventQueue<IVec2>;

pub struct PreviewSystem {
	//layers: Vec<PreviewSystemLayer>,
	database: Arc<Mutex<Database>>,

	layers: Vec<PreviewSystemLayer>,

	notifier: Arc<Notify>,
	pub update_queue_cache: PreviewSystemQueuedChunks,
}

const fn layer_index_to_zoom(index: u8) -> u8 {
	index + 1
}

fn init_layers_vec() -> Vec<PreviewSystemLayer> {
	let mut layers: Vec<PreviewSystemLayer> = Vec::new();
	for i in 0..limits::PREVIEW_SYSTEM_LAYER_COUNT {
		layers.push(PreviewSystemLayer::new(layer_index_to_zoom(i)));
	}
	layers
}

impl PreviewSystem {
	pub fn new(database: Arc<Mutex<Database>>) -> Self {
		let notifier = Arc::new(Notify::new());

		Self {
			database,
			notifier: notifier.clone(),
			update_queue_cache: PreviewSystemQueuedChunks::new(notifier),
			layers: init_layers_vec(),
		}
	}

	const fn layer_index_to_zoom(index: u8) -> u8 {
		index + 1
	}

	pub async fn request_data(&self, pos: &IVec2, zoom: u8) -> anyhow::Result<Option<Arc<Vec<u8>>>> {
		if let Some(record) = DatabaseFunc::preview_load_data(&self.database, *pos, zoom).await? {
			Ok(Some(Arc::new(record.data)))
		} else {
			Ok(None)
		}
	}

	fn process_queue(&mut self) {
		for pos in &self.update_queue_cache.read_all() {
			self.layers[0].add_to_queue(*pos);
		}
	}

	async fn process_layers(preview_system_weak: PreviewSystemWeak) -> anyhow::Result<()> {
		let preview_system_mtx = preview_system_weak
			.upgrade()
			.ok_or_else(|| anyhow!("Preview system expired"))?;
		let mut preview_system = preview_system_mtx.lock().await;
		let mut layers = std::mem::take(&mut preview_system.layers);
		preview_system.layers = init_layers_vec();
		let db_weak = Arc::downgrade(&preview_system.database);
		drop(preview_system);
		drop(preview_system_mtx);

		for i in 0..layers.len() {
			loop {
				if let Some(db) = db_weak.upgrade() {
					let layer = &mut layers[i];
					let res =
						PreviewSystemLayer::process_one_block(&db, &mut layer.update_queue, layer.zoom).await?;

					if !res.has_more {
						break;
					}

					if i < layers.len() - 1 {
						let upper_layer = &mut layers[i + 1];
						upper_layer.add_to_queue(res.upper_pos);
					}
				} else {
					return Err(anyhow!("Database expired"));
				}
			}
		}

		Ok(())
	}

	pub async fn process_all(preview_system_mtx: PreviewSystemMutex) {
		log::trace!("Processing all pending previews");
		let mut preview_system = preview_system_mtx.lock().await;
		preview_system.process_queue();
		drop(preview_system);
		let weak = Arc::downgrade(&preview_system_mtx);
		drop(preview_system_mtx);

		if let Err(e) = Self::process_layers(weak).await {
			// This shouldn't happen on non-corrupted database anyways
			log::error!("Failed to process previews: {e}");
		}
	}
}

impl Drop for PreviewSystem {
	fn drop(&mut self) {
		log::trace!("Preview system freed");
	}
}

pub type PreviewSystemMutex = Arc<Mutex<PreviewSystem>>;
pub type PreviewSystemWeak = Weak<Mutex<PreviewSystem>>;
